library/std/src/sys/unix/fs.rs

```diff
@@ -489,10 +489,18 @@ impl Iterator for ReadDir {
                     };
                 }
 
+                // Only d_reclen bytes of *entry_ptr are valid, so we can't just copy the
+                // whole thing (#93384).  Instead, copy everything except the name.
+                let entry_bytes = entry_ptr as *const u8;
+                let entry_name = (*entry_ptr).d_name.as_ptr() as *const u8;
```

Comment:
The reason I suggested `ptr::addr_of!` is because things like `as_ptr()` require forming a reference `&self`, but our issue here is that `d_name` may not be fully dereferenceable. Keeping everything as pointers should be safer.

Reply:
Ah fair, I did it this way because the code below was already doing the same thing.  I'll change it.

Reply:
Turns out `addr_of!()` isn't even enough: https://github.com/rust-lang/miri/pull/1981#issuecomment-1048278492

Reply:
Ugh, then what? Get the offset from a fully-sized local dummy, then ptr-add on the real thing?

Reply:
Yeah.  That's apparently what all the `offset_of!()` crates do.  Does libstd already have something like that or should I just open code it?

Reply:
My impression was that such use was the very *purpose* of `addr_of!` / `&raw`, and it seems I wasn't the only one from the discussion on #64490 and #73394. But that was left to the future in [RFC 2582](https://rust-lang.github.io/rfcs/2582-raw-reference-mir-operator.html#other).

It seems really unfortunate here since the pointer at hand is not entirely dangling, just _incomplete_ at the end, beyond the pointer we want to compute. A `MaybeUninit` local feels like such a waste, but should be optimized away. Maybe that can even be done in a true `const` these days?

I'm not aware of any existing `offset_of!`, so open-coding it seems fine, with a comment that it _could_ use `offset_of!` so we can find and update it later. :)

Reply:
`addr_of!` is a weaker from of `&`, but it does nothing about `*`. The code here is UB due to the requirement of `*` to only be used on pointers that point to actual memory (with the required size and alignment given by its type).

It would love to just entirely remove that restriction, but then we would have to change our LLVM codegen to use `getelementptr` instead of `getelementptr inbounds` in some situations.

Reply:
> Maybe that can even be done in a true const these days?

With some unstable features, it can -- see the https://github.com/Gilnaa/memoffset crate

Reply:
> I would love to just entirely remove that restriction, but then we would have to change our LLVM codegen to use getelementptr instead of getelementptr inbounds in some situations.

This is probably worth doing, honestly. I can't imagine the use of `gepi` over `gep` wins us almost anything on modern hardware (IIUC it mostly just allows folding the address arithmetic into a subsequent load), and even if it does, we still would be able to use it in all cases except for with addr_of.

Reply:
Cc https://github.com/rust-lang/unsafe-code-guidelines/issues/319